test/celeritas/field/DormandPrinceStepper.test.cu

```diff
@@ -0,0 +1,318 @@
+    constexpr double delta_chord = 1e-4;
+    constexpr double half = 0.5;
+
+    auto id = (blockIdx.x * blockDim.x + threadIdx.x) / *number_threads;
+
+    auto eval = make_dummy_equation(dormand_prince_dummy_field);
+    Stepper_impl stepper{eval};
+    FieldStepperResult res;
+    auto state = states[id];
+    auto step = initial_step_size;
+
+    for (int i = 0; i < *number_iterations; ++i)
+    {
+        res = run_stepper(stepper,
+                          step,
+                          state,
+                          *number_threads,
+                          &ks[id * 7],
+                          &along_state[id],
+                          &results[id]);
+        auto dchord
+            = detail::distance_chord(state, res.mid_state, res.end_state);
+        step *= max(std::sqrt(delta_chord / dchord), half);
+    }
+    results[id] = res;
+}
+}  // namespace
+
+//---------------------------------------------------------------------------//
+// TESTING INTERFACE
+//---------------------------------------------------------------------------//
+//! Run on device and return results
+KernelResult simulate_multi_next_chord(int number_threads,
+                                       int number_states,
+                                       bool use_shared)
+{
+    KernelResult result;
+    int arguments_size = 0;
+
+    // Load initial states and results to device
+    int *d_number_iterations, *d_number_threads;
+    arguments_size += sizeof(int) * 2;
+
+    FieldStepperResult *h_results, *d_results;
+    h_results = new FieldStepperResult[number_states];
+    for (int i = 0; i < number_states; ++i)
+    {
+        h_results[i] = FieldStepperResult();
+        arguments_size += sizeof(FieldStepperResult);
+    }
+
+    OdeState *h_along_state, *d_along_state;
+    if (number_threads > 1 && !use_shared)
+    {
+        h_along_state = new OdeState[number_states];
+        for (int i = 0; i < number_states; ++i)
+        {
+            h_along_state[i] = OdeState();
+            arguments_size += sizeof(OdeState);
+        }
+    }
+
+    OdeState *h_ks, *d_ks;
+    if (number_threads > 1 && !use_shared)
+    {
+        h_ks = new OdeState[number_states * 7];
+        for (int i = 0; i < number_states * 7; ++i)
+        {
+            h_ks[i] = OdeState();
+            arguments_size += sizeof(OdeState);
+        }
+    }
+
+    OdeState *h_states, *d_states;
+    h_states = new OdeState[number_states];
+    for (int i = 0; i < number_states; ++i)
+    {
+        h_states[i] = initial_states[i % number_states_sample];
+        arguments_size += sizeof(OdeState);
+    }
+
+    // Create events
+    cudaEvent_t start, stop;
+    cudaEventCreate(&start);
+    cudaEventCreate(&stop);
+
+    // Allocate memory on device
+    cudaMalloc(&d_results, number_states * sizeof(FieldStepperResult));
+    cudaMalloc(&d_states, number_states * sizeof(OdeState));
+    cudaMalloc(&d_number_iterations, sizeof(int));
```

Comment:
you don't really need to allocate small (less than pointer size?) variables to global memory, it's faster and clearer to let CUDA do a copy and have the variable in register.